### PR TITLE
add webpack to documentation

### PR DESCRIPTION
**Pull request by IdanCo**
## Summary
Add a guide to the official docs on how to integrate Bootstrap with Webpack.

## Motivation
- Webpack's popularity is growing, but there's a lack of reliable up-to-date resources on how to integrate it with Bootstrap
- Following @mdo question in #22196 

## Background
During the past few months i've been integrating bootstrap into several new and existing projects using webpack version 2. This led me to conduct a thorough research and tests on the best practices to achieve this integration. The results of these efforts are in this guide.

## Technical
Check out a sample project which implements the steps detailed in this guide -
https://github.com/IdanCo/webpack-modular/tree/bootstrap4

### CODE REVIEW

**Review by mdo — APPROVED** [review]
Looking at this, the only question I have is around the placement of this on a separate page—need to think about that a bit more. Also, is there anyone else out there using Webpack that would like to chime in with a review here?

File: docs/getting-started/webpack.md
Comment (ajacksified) [inline docs/getting-started/webpack.md]: Since you include examples elsewhere, may be nice to do the same here for clarity:

```
To implement the powerful [options](/getting-started/options/) Bootstrap offers, create a file for variables, such as `_custom.scss`, and import it **before** Bootstrap:

{% highlight scss %}
// theme.scss
$brand-primary: $green;
{% endhighlight %}

{% highlight scss %}
@import "theme.scss";
@import "~bootstrap/scss/bootstrap";
{% endhighlight %}
```

File: docs/getting-started/webpack.md
Comment (ajacksified) [inline docs/getting-started/webpack.md]: May be worth mentioning that this is for webpack v2.

**Review by ajacksified — APPROVED** [review]
I have a few minor optional comments, but otherwise this very similar to my own webpack config.

File: docs/getting-started/webpack.md
Comment (IdanCo) [inline docs/getting-started/webpack.md]: @ajacksified thanks for the feedback! i'll push a new PR later

**Comment by IdanCo** [comment]
@ajacksified , i've explicitly mentioned webpack v2 and put more emphasize on custom variables (after all, why even use webpack if you're not planning on customizing?)

**Comment by ajacksified** [comment]
Sorry! Must have missed it.

**Comment by ajacksified** [comment]
Disregard that, misread the notification email. Looks good to me!

**Comment by samsch** [comment]
I would suggest using [ExtractTextWebpackPlugin](https://webpack.js.org/plugins/extract-text-webpack-plugin/) in the example, rather than style-loader. With style-loader, the styles are injected into `<style>` tags in the html head. With ExtractTextWebpackPlugin, the styles are extracted into a separate css file to be `<link>`ed in.

The downside is the slightly added complexity, but the upside is less "magic" around importing styles from JS.

**Comment by IdanCo** [comment]
@samsch , I agree that using ExtractText in many cases would be the best practice. It's actually used it in the demo project I attached to this PR - https://github.com/IdanCo/webpack-modular/tree/bootstrap4

But I'm not convinced it should be a part of the documentation. There are other best-practices (such as using source maps), but adding them to the documentation might imply that they are **essential** for compiling Bootstrap, even though they are in no way such. Also, i can think of many use cases in which using ExtractText plugin would be counter productive, such as serving dev server or running tests.

I hope other resources (blogs, stackoverflow, etc.) will expand and elaborate this basic setup. I know I plan to...

**Comment by samsch** [comment]
Just a quick counter argument. I'm actually less concerned with ExtractText
being a best practice, and more interested in having a separate css file to
be more similar to environments which users may be used to.

I do see your point about it seeming to be "required" when it is not.
Perhaps that could be satisfied with an additional sentence or two in the
guide noting that it's optional.

On Apr 24, 2017 4:44 PM, "Idan Cohen" <notifications@github.com> wrote:

@samsch <https://github.com/samsch> , I agree that using ExtractText in
many cases would be the best practice. It's actually used it in the demo
project I attached to this PR - https://github.com/IdanCo/
webpack-modular/tree/bootstrap4

But I'm not convinced it should be a part of the documentation. There are
other best-practices (such as using source maps), but adding them to the
documentation might imply that they are *essential* for compiling
Bootstrap, even though they are in no way such. Also, i can think of many
use cases in which using ExtractText plugin would be counter productive,
such as serving dev server or running tests.

I hope other resources (blogs, stackoverflow, etc.) will expand and
elaborate this basic setup. I know I plan to...

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/twbs/bootstrap/pull/22423#issuecomment-296815953>, or mute
the thread

**Comment by mdo** [comment]
Any other feedback before we merge this? (Also, hi @ajacksified! Long time no see! :D)

**Comment by Johann-S** [comment]
My only concern about this PR is, is it the role of BS to show how to integrate BS in any plugins, module bundler etc... ? 
I know Webpack have a huge community and I love this tool but if we do that we'll have to do for others plugins too so we should create a dedicated page for that on our documentation

**Comment by mdo** [comment]
> My only concern about this PR is, is it the role of BS to show how to integrate BS in any plugins, module bundler etc... ?

I think so, yes. I'd love more in-depth and helpful guides for using, extending, and contributing to Bootstrap—regardless of the package manager or community. Bootstrap is meant to be accessible to everyone, beginner or seasoned veteran, so I'm a fan :).

**Comment by IdanCo** [comment]
@mdo totally agree.

@Johann-S i see your point - Bootstrap can't be held responsible for what other bundlers are doing. But on the other hand, the bundlers also can't commit to how Bootstrap is structured (especially while in alfa). So the outcome is that integrations fall "between the chairs". So taking responsibility for documenting may not be the obvious choice, but it will help bring BS to more people, and i think that's worth it.

Regarding which bundlers should be included, IMO as long as there's a demand for integration documentation, it's worth considering adding.

**Comment by Johann-S** [comment]
That's ok for me 👍 it's just a choice to do

**Comment by cr101** [comment]
We are integrating Bootstrap into our web app with [Create React App](https://github.com/facebookincubator/create-react-app)

**Comment by Johann-S** [comment]
I think we can add this page on : Write a **Learn & Extend** or **Approach** section.

Related to #17021 (docs section)

**Comment by Johann-S** [comment]
@IdanCo can you update your branch ? because now we use Popper instead of Tether

**Comment by IdanCo** [comment]
Done! and congrats on the huge Popper PR @Johann-S , can't wait to give it a try!

**Comment by Johann-S** [comment]
Thank you very much @IdanCo ❤️

File: docs/getting-started/webpack.md
Comment (Johann-S) [inline docs/getting-started/webpack.md]: Can you change that by `Popper` ?

File: docs/getting-started/webpack.md
Comment (Johann-S) [inline docs/getting-started/webpack.md]: Can you change that by `Popper` ?

**Review by Johann-S — CHANGES_REQUESTED** [review]

File: docs/getting-started/webpack.md
Comment (IdanCo) [inline docs/getting-started/webpack.md]: On it

**Review by Johann-S — APPROVED** [review]

**Comment by Johann-S** [comment]
LGTM 👍 

The merge is up to @mdo because he said : 
> Looking at this, the only question I have is around the placement of this on a separate page—need to think about that a bit more.

**Comment by IdanCo** [comment]
Changed the webpack configuration to accommodate popper instead of tether.

@Johann-S , any reason why [bootstrap.js](https://github.com/twbs/bootstrap/blob/v4-dev/dist/js/bootstrap.js) in v4-dev is not yet updated and still dependent on thether? I built it manually to do my tests and find the correct recipe, but I'm curious about the repo build process...

BTW, at first glance the new tooltips/popovers/dropdown looks great!

**Comment by IdanCo** [comment]
> Looking at this, the only question I have is around the placement of this on a separate page—need to think about that a bit more.

no problem, let me know if there's anything i can do to help

**Comment by Johann-S** [comment]
Yes we build manualy `dist` files, currently they are outdated but we will update them soon 👍

File: docs/getting-started/webpack.md
Comment (micky2be) [inline docs/getting-started/webpack.md]: Why don't you use the triple back tick?
```js
import 'bootstrap';
```

Comment (mdo) [inline docs/getting-started/webpack.md]: Because we use `highlight` everywhere else due to our Jekyll setup.

**Comment by mdo** [comment]
Thanks y'all!

**Comment by petrpacas** [comment]
@mdo I'd like to point out an issue of compiling the JS with Bootstrap's individual `/dist/` files (as hinted [here](https://github.com/shakacode/bootstrap-loader/issues/172#issuecomment-247164418)).

I tested @IdanCo's repo and it looks fine when using the `bootstrap.js` scripts bundle.

However, if you include the `/dist/` files individually (and necessarily reference them with `ProvidePlugin`), the code is duplicated in the resulting JS file (only referencing them in `ProvidePlugin` and not importing them doesn't work).

**The solution** for individual imports is, as suggested in the first link, to include the individual `/src/` files instead and use a transpiler (Babel etc.).

**Comment by IdanCo** [comment]
Thanks @petrpacas , I'll have a look at it this weekend

**Comment by mselerin** [comment]
Hi @petrpacas. Personally, I don't use the Provide Plugin but the Expose Loader instead (see my comment [here](https://github.com/twbs/bootstrap/issues/22196#issuecomment-288321383))

Maybe you could adapt your import with this loader and see what happen ?

**Comment by petrpacas** [comment]
For those interested, here's my repo with working example of using the individual source files w/ Babel instead:

https://github.com/petrpacas/webpack-bootstrap-4-setup

**Comment by ccalvarez** [comment]
Hi.  

I followed the [guide](https://v4.bootcss.com/docs/4.0/getting-started/webpack/) and for some reason it kept telling me that Tether was needed.  I had to replace Popper with Tether.

I've installed Bootstrap with `npm install --save bootstrap@4.0.0-alpha.6` and in my package.json the dependency is `"bootstrap": "^4.0.0-alpha.6",`.

How can I get the dependency to Popper as explained in the documentation?  
Thanks.

**Comment by IdanCo** [comment]
@ccalvarez, These instructions refer to the develop branch which replaced _Tether_ with _Popper_.

For version Alpha 6 (which you probably use), You can browse [the previous commit](https://raw.githubusercontent.com/twbs/bootstrap/4a907b32573fc9b4b37cacab9538a380d80b5bbc/docs/getting-started/webpack.md), or have a look [in the issue discussion](https://github.com/twbs/bootstrap/issues/22196).

**Comment by ccalvarez** [comment]
@IdanCo I understand, thank you very much!

**Comment by chrisalexander55** [comment]
Although the documentation is very good, I still found it incomplete regarding the actual NPM packages that need to be installed and interop with other loader(s) who generate source-maps. Here is a Gist that brings it together:

[https://gist.github.com/chrisalexander55/81f3d5db057fbcbd74d62d488adbca8a](https://gist.github.com/chrisalexander55/81f3d5db057fbcbd74d62d488adbca8a)

Please feel free to use this to update the docs page.

**Pull request merged**
